store/sqlstore/channel_member_history_store.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package sqlstore
+
+import (
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/store"
+
+	l4g "github.com/alecthomas/log4go"
+)
+
+type SqlChannelMemberHistoryStore struct {
+	SqlStore
+}
+
+func NewSqlChannelMemberHistoryStore(sqlStore SqlStore) store.ChannelMemberHistoryStore {
+	s := &SqlChannelMemberHistoryStore{
+		SqlStore: sqlStore,
+	}
+
+	for _, db := range sqlStore.GetAllConns() {
+		table := db.AddTableWithName(model.ChannelMemberHistory{}, "ChannelMemberHistory").SetKeys(false, "ChannelId", "UserId", "JoinTime")
+		table.ColMap("ChannelId").SetMaxSize(26)
+		table.ColMap("UserId").SetMaxSize(26)
+		table.ColMap("JoinTime").SetNotNull(true)
+	}
+
+	return s
+}
+
+func (s SqlChannelMemberHistoryStore) LogJoinEvent(userId string, channelId string, joinTime int64) store.StoreChannel {
+	return store.Do(func(result *store.StoreResult) {
+		channelMemberHistory := &model.ChannelMemberHistory{
+			UserId:    userId,
+			ChannelId: channelId,
+			JoinTime:  joinTime,
+		}
+
+		if err := s.GetMaster().Insert(channelMemberHistory); err != nil {
+			result.Err = model.NewAppError("SqlChannelMemberHistoryStore.LogJoinEvent", "store.sql_channel_member_history.log_join_event.app_error", map[string]interface{}{"ChannelMemberHistory": channelMemberHistory}, err.Error(), http.StatusInternalServerError)
+		}
+	})
+}
+
+func (s SqlChannelMemberHistoryStore) LogLeaveEvent(userId string, channelId string, leaveTime int64) store.StoreChannel {
+	return store.Do(func(result *store.StoreResult) {
+		query := `
+			SELECT
```

Comment:
This seems a bit complicated. Why not just:
```sql
SELECT * FROM ChannelMemberHistory WHERE UserId = :UserId AND ChannelId = :ChannelId ORDER BY JoinTime ASC LIMIT 1
```
Then we can check the NULL app side.

Also a bit concerned that leaving requires an query and update to this table. If it was just an insert we wouldn't have to worry too much about indexes on this table and effects of this table getting large because only the export would take that perf hit.

Reply:
*Regarding the SELECT:*
We're looking for an existing record that has a join time but no leave time, meaning that the user joined the channel, but hasn't yet left. There should be exactly zero or one of these records for any given user/channel combination. If we take out the LeaveTime clause and handle the NULL in business logic, we might have to iterate over a few hundred results looking for the one we care about. Why not let the database do what databases do well?

*Regarding the Insert/Update Logic:*
True, a straight insert would make this logic simpler, but it would complicate the `GetUsersInChannelDuring(...)` logic, because we'd have to do a reflexive join there to match the join and leave events to one another. Either way, you incur complexity. The advantage of your approach is that the code can always run on a read replica rather than the master, but I haven't seen any evidence that my approach causes a performance problem, and I think it's easier to understand.

The other thing to keep in mind is that if companies are using the Message Export feature, the contents of the ChannelMemberHistory table will be purged after every successful export, so there should be at most thousands of records in the table. It's less clear to me what might happen if a company with millions of users doesn't use the feature for years on end, and the size of this table builds up over time.

Reply:
**Regarding the SELECT**:
Right. If we guarantee 0 or 1 of them, then why do the ORDER BY? We can just do:
```sql
SELECT * FROM ChannelMemberHistory WHERE UserId = :UserId AND ChannelId = :ChannelId AND LeaveTime IS NULL LIMIT 1
```
Also change the JoinTime to LeaveTime on the table index (line 25)

**Regarding the Insert/Update Logic**:

> It's less clear to me what might happen if a company with millions of users doesn't use the feature for years on end, and the size of this table builds up over time.

I think that's my main concern. 

> The advantage of your approach is that the code can always run on a read replica rather than the master

Huh? Mine would only run on master because it's insert only. My concern is about not paying performance cost in little chucks a long the way even if you don't use the feature. I would rather the performance cost be mostly concentrated when you actually export. 

> The other thing to keep in mind is that if companies are using the Message Export feature, the contents of the ChannelMemberHistory table will be purged after every successful export

I didn't realize that the table would ever be purged. So you mean if I test the feature before I use it I am screwed? Or if I slightly mistype my command and, for example, export in the wrong format I can never get that data back? Seems like a usability flaw to me.

Reply:
> The other thing to keep in mind is that if companies are using the Message Export feature, the contents of the ChannelMemberHistory table will be purged after every successful export

Also interested in clarification on this. This seems unnecessarily destructive.

Reply:
Yup, and that behaviour will be removed as a part of https://mattermost.atlassian.net/browse/PLT-8238